src/wasm/wasm.cpp

```diff
@@ -974,8 +958,7 @@ Type BrOn::getCastType() {
       // BrOnNull does not send a value on the branch.
       return Type::none;
     case BrOnCast:
-      // FIXME: When we support non-nullable types, this should be non-nullable.
-      return Type(rtt->type.getHeapType(), Nullable);
+      return Type(rtt->type.getHeapType(), NonNullable);
```

Comment:
You're right here as well, it was me that missed something...

Reply:
Actually I think was correct before. This is the cast type, the type that is sent if the cast succeeds and we branch. That is non-nullable, https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#casts

Reply:
Which part of this does say it is nonnullable? I'm also not sure what `t0` is...
> * `br_on_cast <labelidx>` branches if a value can be cast down to a given reference type
>  - `br_on_cast $l : [t0* t (rtt n? $t')] -> [t0* t]`
>    - iff `$l : [t0* t']`
>    - and `t <: (ref null data)` or `t <: (ref null func)`
>    - and `(ref $t') <: t'`
>  - branches iff the first operand is not null and its runtime type is a sub-RTT of the RTT operand
>  - passes cast operand along with branch, plus possible extra args

Reply:
`t0*` is a sequence of arbitrary types that sent along the branch, just like how `br` can carry any number of types. The key part for the non-nullability is the condition `(ref $t') <: t'` which says that the label is allowed to expect a non-nullable reference type from the branch.

Reply:
Thanks! But some more questions:
- Is `$l : [t0* t (rtt n? $t')] -> [t0* t]` the rule for cast (=sent) type, or the normal return type within the block? If it's one of them, where is the rule for the other?
- Don't understand what `(ref $t') <: t'` means. How can `(ref $t') <: t'` be satisfied for any given type `t'`?

Reply:
The `$l` is the branch label and its type is given by `$l : [t0* t']`. In `br_cast $l`, the `$l` is just given as an immediate operand of the instruction and the type of the instruction in its block is `[t0* t (rtt n? $t')] -> [t0* t]`.

For example, `(ref $t') <: t'` would be true for `t' = (ref $t')`, but also for `t' = (ref null $t')`. (Note that `t'` is a value type but `$t'` is a type index.)

Reply:
> The `$l` is the branch label and its type is given by `$l : [t0* t']`. In `br_cast $l`, the `$l` is just given as an immediate operand of the instruction and the type of the instruction in its block is `[t0* t (rtt n? $t')] -> [t0* t]`.

Sorry not sure if I get it... The same question: Is this rule for the sent type or the type within the block? So it has to kinds of outputs, one when the branch is taken, and the other when it is not taken and falls through. When does `[t0* t (rtt n? $t')] -> [t0* t]` apply? From you explanation it sounds like it is the sent type (when branch is taken), but you also said "type of instruction in its block". And if this is one of them, what is the rule for the other?

> For example, `(ref $t') <: t'` would be true for `t' = (ref $t')`, but also for `t' = (ref null $t')`. (Note that `t'` is a value type but `$t'` is a type index.)

I still don't get what this rule means. Which part of this rule explains the condition for the casting? This rule only contains `t'` and not `t`, then how can this be a rule describing the condition?

---
Also not sure about thie rule:
>   * and `t <: (ref null data)` or `t <: (ref null func)`
Does this mean the input argument should be nullable, and we can't use nonnullable types for `br_on_cast`?

Sorry for many questions ;(

Reply:
The input type to the `br_on_cast` is the same `[t0* t (rtt n? $t')]` in both cases. When it branches, it sends `[t0* t']` to the label. When it doesn't branch, it produces `[t0* t]`. That is, when it branches, the result is a non-null value and when it doesn't branch, the value and its type are unchanged (but the rtt is consumed in either case).

> Which part of this rule explains the condition for the casting? 

This is a restriction on the label (i.e. block) that `br_on_cast` is allowed to branch to. The label must expect to produce some supertype of `(ref $t')` because the `br_on_cast` is going to send it a `(ref $t')` value. We know that `t'` is a subtype of `t` from the text "branches iff the first operand is not null and its runtime type is a sub-RTT of the RTT operand," which refers to `t` and `t'`, although not by name. The "sub-RTT" relationship requires that `t'` is a subtype of `t`.

>> and `t <: (ref null data)` or t <: `(ref null func)`
>
> Does this mean the input argument should be nullable, and we can't use nonnullable types for br_on_cast?

It means that the input _may_ be nullable, but doesn't have to be. Remember that non-nullable types are subtypes of their nullable counterparts, so they can be used wherever the nullable types are allowed.

Reply:
I see. Thanks for the detailed explanation!